src/grid/cubic.py

```diff
@@ -0,0 +1,516 @@
+        def x_spline(x, y, z, nu_x):
+            # x, y, z may not be in the grid.
+            val = CubicSpline(
+                self.points[np.arange(1, self.num_pts[0] - 2) * self.num_pts[1] * self.num_pts[2], 0],
+                [y_splines(y, x_index, z, nu_y) for x_index in range(1, self.num_pts[0] - 2)],
+            )(x, nu_x)
+
+            return val
+
+        print(sum([nu_x == 0, nu_y == 0, nu_z == 0]) == 2, 1 in (nu_x, nu_y, nu_z))
+        # Only consider taking the derivative in only one direction.
+        one_var_deriv = sum([nu_x == 0, nu_y == 0, nu_z == 0]) == 2 and 1 in (nu_x, nu_y, nu_z)
+        interpolated = x_spline(real_pt[0], real_pt[1], real_pt[2], nu_x)
+        if use_log and (nu_x, nu_y, nu_z) == (0, 0, 0):
+            return np.exp(interpolated)
+        elif use_log and one_var_deriv:
+            interpolated_func = self.interpolate_function(real_pt, func_values, use_log=False)
+            return interpolated * np.exp(interpolated_func)
+        if use_log and sum([nu_x == 0, nu_y == 0, nu_z == 0]) == 2:
+            if nu_x > 0:
+                derivs = [
+                    self.interpolate_function(real_pt, func_values, use_log=False,
+                                              nu_x=i, nu_y=0, nu_z=0) for i in range(0, nu_x)
+                ]
+            elif nu_y > 0:
+                derivs = [
+                    self.interpolate_function(real_pt, func_values, use_log=False,
+                                              nu_x=0, nu_y=i, nu_z=0) for i in range(0, nu_y)
+                ]
+            else:
+                derivs = [
+                    self.interpolate_function(real_pt, func_values, use_log=False,
+                                              nu_x=0, nu_y=0, nu_z=i) for i in range(0, nu_z)
+                ]
+            return derivs[0] * Bell_polynomial(*derivs[1:])
+        return interpolated
+
+    def _indices_to_index(self, indices):
+        r"""
+        Convert Indices to Index, ie (i, j, k) to a index/integer m corresponding to grid point.
+
+        Parameters
+        ----------
+        indices : (int, int, int)
+            The ith, jth, kth position of the grid point.
+
+        Returns
+        -------
+        index : int
+            Index of the grid point.
+
+        """
+        n_1d, n_2d = self.num_pts[2], self.num_pts[1] * self.num_pts[2]
+        index = n_2d * indices[0] + n_1d * indices[1] + indices[2]
+        return index
+
+
+class UniformCubicGrid(CubicGrid):
+    def __init__(self, l_bnd, u_bnd, step_sizes):
+        """
+        Construct the UniformCubicGrid object.
+
+        Parameters
+        ----------
+        smallest_pnt : ndarray(3,)
+            The smallest point in each axis in the 3D cubic grid.
+        largest_pnt : ndarray(3,)
+            The largest point in each axis in the 3D cubic grid.
+        step_sizes : ndarray(3,)
+            The step-size between two consecutive points in each axis in the 3D cubic grid.
```

Comment:
Usually, the user specifies the origin (or the `smallest_pnt`), the `step_size`, and the number of points in each direction. I don't think requiring the `largest_pnt` is that user-friendly.

Reply:
Resolved in d958179